File: qa/georgian-l10n-phase3/_trace_ka.py
# -*- coding: utf-8 -*-
"""Extract added ka.ts dotted keys vs HEAD and trace consumers. No git writes."""
from __future__ import annotations

import json
import re
import subprocess
from pathlib import Path

ROOT = Path(r"c:\Users\User\Desktop\www.medicard")
OUT_DIR = ROOT / "qa" / "georgian-l10n-phase3"
KA = ROOT / "mobile" / "src" / "i18n" / "ka.ts"
MOBILE = ROOT / "mobile"
GEO = re.compile(r"[\u10A0-\u10FF]")
PH = re.compile(r"(\{[A-Za-z0-9_]+\}|\$\{[^}]+\}|%[sdif])")

SEARCH_EXTS = {".ts", ".tsx", ".js", ".jsx"}
SKIP_PARTS = {"node_modules", "dist"}

DIRECT_RE = re.compile(r"(შენ |შენი |შენთვის|გინდა|სცადე[^თ]|შეამოწმე[^თ]|აღრიცხე[^თ]|დაამატე[^თწ]|ჰკითხე)")
FORMAL_RE = re.compile(r"(თქვენ|თქვენი|გსურთ|სცადეთ|შეამოწმეთ|აირჩიეთ|მიუთითეთ|დაამატეთ|შეინახეთ|დალიეთ)")


def extract_keys(src: str) -> dict[str, str]:
    """Walk object literals after `export const ka = {` for string/template values."""
    start = src.find("export const ka")
    if start < 0:
        start = 0
    i = src.find("{", start)
    keys: dict[str, str] = {}
    path: list[str] = []
    n = len(src)

    def skip_ws(p):
        while p < n and src[p] in " \t\r\n":
            p += 1
        return p

    def read_ident(p):
        m = re.match(r"[A-Za-z_][A-Za-z0-9_]*", src[p:])
        if not m:
            return None, p
        return m.group(0), p + m.end()

    def read_string(p):
        q = src[p]
        if q not in "'\"`":
            return None, p
        p += 1
        out = []
        while p < n:
            ch = src[p]
            if ch == "\\" and p + 1 < n:
                out.append(src[p : p + 2])
                p += 2
                continue
            if ch == q:
                return "".join(out), p + 1
            out.append(ch)
            p += 1
        return "".join(out), p

    p = i + 1
    expecting_key = True
    while p < n:
        p = skip_ws(p)
        if p >= n:
            break
        if src.startswith("//", p):
            p = src.find("\n", p)
            if p < 0:
                break
            p += 1
            continue
        if src.startswith("/*", p):
            end = src.find("*/", p + 2)
            p = n if end < 0 else end + 2
            continue
        if src[p] == "}":
            if path:
                path.pop()
            p += 1
            p = skip_ws(p)
            if p < n and src[p] == ",":
                p += 1
            expecting_key = True
            if not path and p > i + 1:
                # closed root
                break
            continue
        if not expecting_key:
            # skip until comma or brace handled
            if src[p] == ",":
                p += 1
                expecting_key = True
                continue
            p += 1
            continue

        ident, np = read_ident(p)
        if ident is None:
            if src[p] in "'\"":
                ident, np = read_string(p)
            elif src[p] == "[":
                # skip computed
                p += 1
                continue
            else:
                p += 1
                continue
        p = skip_ws(np)
        if p < n and src[p] == "?":
            p += 1
            p = skip_ws(p)
        if p >= n or src[p] != ":":
            p = np
            continue
        p = skip_ws(p + 1)

        # function
        if src.startswith("(", p) or src.startswith("async", p):
            # find => then value
            arrow = src.find("=>", p)
            if arrow < 0:
                p += 1
                continue
            p = skip_ws(arrow + 2)
            if src[p] == "{":
                # block function — skip balanced
                depth = 0
                q = p
                while q < n:
                    if src[q] == "{":
                        depth += 1
                    elif src[q] == "}":
                        depth -= 1
                        if depth == 0:
                            q += 1
                            break
                    q += 1
                dotted = ".".join(path + [ident])
                snippet = src[p:q]
                if GEO.search(snippet):
                    keys[dotted] = snippet[:240].replace("\n", " ")
                p = skip_ws(q)
                if p < n and src[p] == ",":
                    p += 1
                expecting_key = True
                continue
            val, p = read_string(p) if src[p] in "'\"`" else (None, p)
            dotted = ".".join(path + [ident])
            if val is not None:
                keys[dotted] = val
            p = skip_ws(p)
            if p < n and src[p] == ",":
                p += 1
            expecting_key = True
            continue

        if src[p] == "{":
            path.append(ident)
            p += 1
            expecting_key = True
            continue

        if src[p] in "'\"`":
            val, p = read_string(p)
            dotted = ".".join(path + [ident])
            keys[dotted] = val or ""
            p = skip_ws(p)
            # type assertions `as const`
            if src.startswith("as ", p):
                semi = src.find("\n", p)
                p = skip_ws(semi if semi > 0 else p)
            if p < n and src[p] == ",":
                p += 1
            expecting_key = True
            continue

        if src[p] == "[":
            # array — capture if Georgian
            depth = 0
            q = p
            while q < n:
                if src[q] == "[":
                    depth += 1
                elif src[q] == "]":
                    depth -= 1
                    if depth == 0:
                        q += 1
                        break
                q += 1
            snippet = src[p:q]
            dotted = ".".join(path + [ident])
            if GEO.search(snippet):
                keys[dotted] = snippet[:240].replace("\n", " ")
            p = skip_ws(q)
            if src.startswith("as ", p):
                nl = src.find("\n", p)
                p = skip_ws(nl if nl > 0 else p)
            if p < n and src[p] == ",":
                p += 1
            expecting_key = True
            continue

        # unknown value (spread, identifier)
        dotted = ".".join(path + [ident])
        if src.startswith("...", p):
            keys[dotted] = "(spread)"
        p += 1
        expecting_key = False

    return keys


def voice_of(text: str, dotted: str) -> str:
    if any(
        x in dotted
        for x in (
            ".title",
            "Title",
            "hubTitle",
            "label",
            "Label",
            "unit",
            "placeholder",
            "Placeholder",
        )
    ) and not FORMAL_RE.search(text) and not DIRECT_RE.search(text):
        kind = "label"
    else:
        kind = "neutral"
    if FORMAL_RE.search(text) and DIRECT_RE.search(text):
        return "mixed"
    if FORMAL_RE.search(text):
        return "formal"
    if DIRECT_RE.search(text):
        return "direct"
    return kind


def collect_source_files() -> list[Path]:
    files = []
    for p in MOBILE.rglob("*"):
        if p.suffix.lower() not in SEARCH_EXTS:
            continue
        if any(part in SKIP_PARTS for part in p.parts):
            continue
        files.append(p)
    return files


SCREEN_HINTS = [
    ("app/(auth)/sign-in", "Authentication / login"),
    ("app/(auth)/sign-up", "Authentication / registration"),
    ("app/(auth)/phone", "Authentication / phone OTP"),
    ("app/(auth)/forgot-password", "Authentication / password recovery"),
    ("app/(auth)/assessment", "Onboarding / assessment"),
    ("app/(auth)/profile-setup", "Onboarding / profile setup"),
    ("app/(tabs)/index", "Home"),
    ("app/chat/", "Medi chat"),
    ("app/cycle/pregnancy", "Pregnancy"),
    ("app/cycle/", "Cycle"),
    ("components/cycle/", "Cycle"),
    ("app/health-metrics/hydration", "Hydration"),
    ("components/hydration", "Hydration"),
    ("components/medications", "Medications"),
    ("app/medications", "Medications"),
    ("medi-quest", "Quest"),
    ("components/quest", "Quest"),
    ("notification", "Notifications"),
    ("pushCopy", "Push / notifications"),
    ("app/(tabs)/profile", "Profile"),
    ("app/profile", "Profile / settings"),
    ("i18n/ka.ts", "catalog definition"),
]


def screen_for(paths: list[str]) -> str:
    joined = " ".join(paths)
    for needle, label in SCREEN_HINTS:
        if needle.replace("\\", "/") in joined.replace("\\", "/"):
            return label
    if not paths:
        return "—"
    return "unresolved reference"


def main():
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    now = KA.read_text(encoding="utf-8")
    old = subprocess.check_output(
        ["git", "show", "HEAD:mobile/src/i18n/ka.ts"],
        cwd=ROOT,
        encoding="utf-8",
        errors="replace",
    )
    old_keys = extract_keys(old)
    new_keys = extract_keys(now)
    added = sorted(set(new_keys) - set(old_keys))
    changed = sorted(k for k in (set(old_keys) & set(new_keys)) if old_keys[k] != new_keys[k])

    sources = collect_source_files()
    blobs: list[tuple[str, str]] = []
    for p in sources:
        rel = p.relative_to(MOBILE).as_posix()
        if rel == "src/i18n/ka.ts":
            continue
        try:
            blobs.append((rel, p.read_text(encoding="utf-8", errors="replace")))
        except Exception:
            continue

    rows = []
    for dotted in added:
        leaf = dotted.split(".")[-1]
        consumers = []
        for rel, text in blobs:
            if f"ka.{dotted}" in text:
                consumers.append(rel)
                continue
            # ka.cycle.foo
            if dotted in text and ("ka." + dotted.split(".")[0]) in text:
                # tighter: ka.section.leaf
                section = dotted.split(".")[0]
                if f"ka.{section}.{leaf}" in text or f"ka.{section}?.{leaf}" in text:
                    consumers.append(rel)
                    continue
            if f"titleKey: '{leaf}'" in text or f'titleKey: "{leaf}"' in text:
                consumers.append(rel)
                continue
            if f"bodyKey: '{leaf}'" in text or f'bodyKey: "{leaf}"' in text:
                consumers.append(rel)
                continue
            if f"copy.{leaf}" in text or f"ka.cycle.{leaf}" in text:
                consumers.append(rel)

        consumers = sorted(set(consumers))
        text = new_keys.get(dotted, "")
        phs = ",".join(PH.findall(text))
        if not consumers:
            status = "unused"
        elif any("push" in c.lower() or "notification" in c.lower() for c in consumers):
            status = "verified in push/backend usage"
        else:
            status = "verified in component"
        # generated: functions
        if "(" in text[:20] and "=>" in text:
            if status == "unused":
                status = "generated dynamically"
        rows.append(
            {
                "key": dotted,
                "consumers": consumers,
                "screen": screen_for(consumers),
                "voice": voice_of(text, dotted),
                "placeholders": phs,
                "status": status,
                "preview": text[:160],
            }
        )

    md = []
    md.append("# Phase 3 — added `ka.ts` key traceability")
    md.append("")
    md.append(f"HEAD keys parsed: **{len(old_keys)}**")
    md.append(f"Worktree keys parsed: **{len(new_keys)}**")
    md.append(f"Added keys: **{len(added)}**")
    md.append(f"Changed existing keys (value differs): **{len(changed)}**")
    md.append("")
    counts = {}
    for r in rows:
        counts[r["status"]] = counts.get(r["status"], 0) + 1
    md.append("## Status counts")
    for k, v in sorted(counts.items()):
        md.append(f"- {k}: **{v}**")
    md.append("")
    md.append("| Key | Consumer file(s) | Screen/state | Voice | Placeholder(s) | Status |")
    md.append("| --- | ---------------- | ------------ | ----- | -------------- | ------ |")
    for r in rows:
        cons = "<br>".join(r["consumers"]) if r["consumers"] else "—"
        md.append(
            f"| `{r['key']}` | {cons} | {r['screen']} | {r['voice']} | `{r['placeholders'] or '—'}` | {r['status']} |"
        )

    unused = [r for r in rows if r["status"] == "unused"]
    md.append("")
    md.append("## Unused added keys")
    md.append(f"Count: **{len(unused)}**. Not deleted.")
    for r in unused:
        md.append(f"- `{r['key']}` — {r['preview'][:80]}")

    (OUT_DIR / "ka-added-keys.md").write_text("\n".join(md), encoding="utf-8")
    (OUT_DIR / "ka-added-keys.json").write_text(
        json.dumps(
            {
                "head_key_count": len(old_keys),
                "worktree_key_count": len(new_keys),
                "added_count": len(added),
                "changed_existing_count": len(changed),
                "status_counts": counts,
                "rows": rows,
            },
            ensure_ascii=False,
            indent=2,
        ),
        encoding="utf-8",
    )
    print("added", len(added), "changed_existing", len(changed))
    print("status", counts)
    print("wrote", OUT_DIR / "ka-added-keys.md")


if __name__ == "__main__":
    main()

File: qa/georgian-l10n-phase3/_trace_ka.mjs
/**
 * Trace ka.ts keys added vs git HEAD. Read-only except writing this QA folder.
 */
import { execSync } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import { createRequire } from 'node:module';
import { fileURLToPath } from 'node:url';
import ts from '../../mobile/node_modules/typescript/lib/typescript.js';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..', '..');
const OUT = path.join(ROOT, 'qa', 'georgian-l10n-phase3');
const KA = path.join(ROOT, 'mobile', 'src', 'i18n', 'ka.ts');
const require = createRequire(import.meta.url);
const GEO = /[\u10A0-\u10FF]/;
const PH = /(\{[A-Za-z0-9_]+\}|\$\{[^}]+\}|%[sdif])/g;
const DIRECT_RE =
  /(შენ |შენი |შენთვის|\bგინდა\b|სცადე[^თ]|შეამოწმე[^თ]|აღრიცხე[^თ]|დაამატე[^თწნ]|ჰკითხე|დალიე[^თ])/;
const FORMAL_RE =
  /(თქვენ|თქვენი|გსურთ|სცადეთ|შეამოწმეთ|აირჩიეთ|მიუთითეთ|დაამატეთ|შეინახეთ|დალიეთ|შეიყვანეთ)/;

function transpileKa(src, extraPrelude = '') {
  const stripped = src.replace(/^import[\s\S]*?from ['"][^'"]+['"];?\r?\n/gm, '');
  const { outputText } = ts.transpileModule(stripped, {
    compilerOptions: {
      module: ts.ModuleKind.CommonJS,
      target: ts.ScriptTarget.ES2020,
      strict: false,
    },
    fileName: 'ka.ts',
  });
  const prelude = extraPrelude + '\n';
  const wrapped = `${prelude}${outputText}`;
  const m = { exports: {} };
  const fn = new Function('module', 'exports', 'require', wrapped);
  fn(m, m.exports, require);
  return m.exports.ka;
}

function flatten(obj, prefix = '', out = {}) {
  if (obj == null) {
    if (prefix) out[prefix] = String(obj);
    return out;
  }
  if (typeof obj === 'function') {
    out[prefix] = obj.toString();
    return out;
  }
  if (typeof obj !== 'object') {
    out[prefix] = String(obj);
    return out;
  }
  if (Array.isArray(obj)) {
    const geo = obj.some((x) => typeof x === 'string' && GEO.test(x));
    if (geo || obj.every((x) => typeof x !== 'object' || x == null)) {
      out[prefix] = JSON.stringify(obj);
      return out;
    }
    obj.forEach((item, i) => flatten(item, `${prefix}.${i}`, out));
    return out;
  }
  for (const [k, v] of Object.entries(obj)) {
    flatten(v, prefix ? `${prefix}.${k}` : k, out);
  }
  return out;
}

function voiceOf(text, dotted) {
  const t = String(text || '');
  const hasF = FORMAL_RE.test(t);
  const hasD = DIRECT_RE.test(t);
  if (hasF && hasD) return 'mixed';
  if (hasF) return 'formal';
  if (hasD) return 'direct';
  const labelish = /(title|Title|label|Label|placeholder|Placeholder|unit|hub|name|Name)$/.test(
    dotted.split('.').pop() || '',
  );
  return labelish ? 'label' : 'neutral';
}

function walkFiles(dir, acc = []) {
  let entries = [];
  try {
    entries = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return acc;
  }
  for (const e of entries) {
    const p = path.join(dir, e.name);
    if (e.isDirectory()) {
      if (['node_modules', 'dist', '.git'].includes(e.name)) continue;
      walkFiles(p, acc);
    } else if (/\.(ts|tsx|js|jsx)$/.test(e.name)) {
      acc.push(p);
    }
  }
  return acc;
}

const SCREEN_HINTS = [
  ['app/(auth)/sign-in', 'Authentication / login'],
  ['app/(auth)/sign-up', 'Authentication / registration'],
  ['app/(auth)/phone', 'Authentication / phone OTP'],
  ['app/(auth)/forgot-password', 'Authentication / password recovery'],
  ['app/(auth)/assessment', 'Onboarding / assessment'],
  ['app/(auth)/profile-setup', 'Onboarding / profile setup'],
  ['app/(tabs)/home', 'Home'],
  ['app/(tabs)/index', 'Home'],
  ['components/home/', 'Home'],
  ['app/chat/', 'Medi chat'],
  ['components/chat/', 'Medi chat'],
  ['app/cycle/pregnancy', 'Pregnancy'],
  ['pregnancyCare', 'Pregnancy / care planner'],
  ['pregnancyTimeline', 'Pregnancy / timeline'],
  ['app/cycle/', 'Cycle'],
  ['components/cycle/', 'Cycle'],
  ['health-metrics/hydration', 'Hydration'],
  ['components/hydration', 'Hydration'],
  ['lib/hydration', 'Hydration'],
  ['components/medications', 'Medications'],
  ['app/medications', 'Medications'],
  ['app/(tabs)/meds', 'Medications'],
  ['medi-quest', 'Quest'],
  ['components/quest', 'Quest'],
  ['i18n/quest', 'Quest'],
  ['notification', 'Notifications'],
  ['pushCopy', 'Push / notifications'],
  ['pushTemplates', 'Push / backend'],
  ['app/(tabs)/profile', 'Profile'],
  ['app/profile', 'Profile / settings'],
  ['lib/companion', 'Medi companion'],
  ['lib/weather', 'Home / weather'],
  ['constants/cycle', 'Cycle constants'],
  ['constants/conditionCatalog', 'Health profile / conditions'],
];

function screenFor(paths) {
  const joined = paths.join(' ').replaceAll('\\', '/');
  for (const [needle, label] of SCREEN_HINTS) {
    if (joined.includes(needle)) return label;
  }
  return paths.length ? 'unresolved reference' : '—';
}

function loadSources() {
  const roots = [path.join(ROOT, 'mobile'), path.join(ROOT, 'server', 'src')];
  const blobs = [];
  for (const root of roots) {
    for (const file of walkFiles(root)) {
      const rel = path.relative(ROOT, file).replaceAll('\\', '/');
      if (rel === 'mobile/src/i18n/ka.ts') continue;
      if (rel.endsWith('_trace_ka.mjs')) continue;
      try {
        blobs.push({ rel, text: fs.readFileSync(file, 'utf8') });
      } catch {
        /* ignore */
      }
    }
  }
  return blobs;
}

const DYNAMIC_PARENTS = new Set([
  'periTrendFamily',
  'pregnancyTrendFamily',
  'periTrendFlow',
  'pregnancyTrendFlow',
  'energyLevel',
  'trendGroup',
  'painType',
  'painSeverity',
  'formLabels',
  'dosageUnitLabels',
  'mealTiming',
]);

function findConsumers(dotted, blobs) {
  const parts = dotted.split('.');
  const leaf = parts[parts.length - 1];
  const parent = parts.length >= 2 ? parts[parts.length - 2] : '';
  const section = parts[0];
  const hits = [];
  const shortLeaf = leaf.length < 6 || /^(title|body|label|hint|empty|cta|open|add|save|note)$/i.test(leaf);

  for (const { rel, text } of blobs) {
    if (text.includes(`ka.${dotted}`)) {
      hits.push(rel);
      continue;
    }
    if (parts.length === 2 && text.includes(`ka.${section}.${leaf}`)) {
      hits.push(rel);
      continue;
    }
    if (dotted.startsWith('cycle.carePlan.')) {
      if (text.includes('ka.cycle.carePlan') && text.includes(`copy.${leaf}`)) {
        hits.push(rel);
        continue;
      }
      if (text.includes(`pregnancyCareCopy('${leaf}')`) || text.includes(`pregnancyCareCopy("${leaf}")`)) {
        hits.push(rel);
        continue;
      }
    }
    if (dotted.startsWith('cycle.timeline.')) {
      if (text.includes('ka.cycle.timeline') && text.includes(`copy.${leaf}`)) {
        hits.push(rel);
        continue;
      }
      if (text.includes('timelineCopy(') && (text.includes(`'${leaf}'`) || text.includes(`"${leaf}"`))) {
        hits.push(rel);
        continue;
      }
    }
    if (parent && !shortLeaf && text.includes(`ka.${section}.${parent}`)) {
      hits.push(rel);
      continue;
    }
    if (
      !shortLeaf &&
      (text.includes(`'${leaf}'`) || text.includes(`"${leaf}"`) || text.includes(`.${leaf}`))
    ) {
      if (
        text.includes(`ka.${section}`) ||
        (section === 'cycle' && text.includes('ka.cycle'))
      ) {
        hits.push(rel);
      }
    }
  }
  return [...new Set(hits)];
}

function classify(dotted, consumers, preview) {
  const parts = dotted.split('.');
  const leaf = parts[parts.length - 1];
  const parent = parts[parts.length - 2] || '';
  const uiConsumers = consumers.filter((c) => /\/(app|components)\//.test(c));
  const catalogOnly = consumers.length > 0 && consumers.every((c) => /i18n\//.test(c));

  if (DYNAMIC_PARENTS.has(parent) || /^hydration\.levels\./.test(dotted) || /^assessment\.steps\./.test(dotted)) {
    return 'generated dynamically';
  }
  if (/^cycle\.(carePlan|timeline)\.(care_|ms_|src_)/.test(dotted)) {
    return 'generated dynamically';
  }
  if (/^cycle\.timeline\.(past|current|upcoming)$/.test(dotted)) {
    return 'generated dynamically';
  }
  if (consumers.some((c) => /pushCopy|pushTemplates|notificationCatalog|notificationPlan/.test(c))) {
    return 'verified in push/backend usage';
  }
  if (uiConsumers.length) return 'verified in component';
  if (catalogOnly) return 'generated dynamically';
  if (!consumers.length) {
    if (typeof preview === 'string' && preview.includes('=>')) return 'generated dynamically';
    return 'unused';
  }
  if (consumers.some((c) => /\.(tsx|ts|js)$/.test(c) && !/i18n\//.test(c))) {
    return 'verified in component';
  }
  return 'unresolved reference';
}

function placeholdersOf(text) {
  const s = String(text || '');
  const found = s.match(PH) || [];
  // also capture template ${} inside function source
  return [...new Set(found)].join(', ');
}

function duplicateOf(dotted, preview, all) {
  const norm = String(preview).replace(/\s+/g, ' ').trim();
  if (norm.length < 12) return '';
  for (const [k, v] of Object.entries(all)) {
    if (k === dotted) continue;
    if (String(v).replace(/\s+/g, ' ').trim() === norm) return k;
  }
  return '';
}

const nowSrc = fs.readFileSync(KA, 'utf8');
const oldSrc = execSync('git show HEAD:mobile/src/i18n/ka.ts', { cwd: ROOT, encoding: 'utf8' });

const pregnancyCare = require(path.join(ROOT, 'mobile', 'src', 'i18n', 'cycle', 'pregnancyCare.js'));
const pregnancyTimeline = require(path.join(ROOT, 'mobile', 'src', 'i18n', 'cycle', 'pregnancyTimeline.js'));
const prelude = `
const { PREGNANCY_CARE_COPY_KA } = require(${JSON.stringify(path.join(ROOT, 'mobile', 'src', 'i18n', 'cycle', 'pregnancyCare.js'))});
const { PREGNANCY_TIMELINE_COPY_KA } = require(${JSON.stringify(path.join(ROOT, 'mobile', 'src', 'i18n', 'cycle', 'pregnancyTimeline.js'))});
`;

const nowKa = transpileKa(nowSrc, prelude);
const oldKa = transpileKa(oldSrc, '');
const nowFlat = flatten(nowKa);
const oldFlat = flatten(oldKa);
const added = Object.keys(nowFlat).filter((k) => !(k in oldFlat)).sort();
const changed = Object.keys(nowFlat)
  .filter((k) => k in oldFlat && String(nowFlat[k]) !== String(oldFlat[k]))
  .sort();

const blobs = loadSources();
const rows = added.map((key) => {
  const preview = nowFlat[key];
  const consumers = findConsumers(key, blobs);
  let status = classify(key, consumers, preview);
  // hydration numeric levels are indexed dynamically
  if (key.startsWith('hydration.levels.') && consumers.length === 0) {
    const hydra = blobs.filter((b) => /hydration/i.test(b.rel) && /levels/.test(b.text));
    if (hydra.length) {
      consumers.push(...hydra.map((h) => h.rel));
      status = 'generated dynamically';
    }
  }
  if (key.startsWith('assessment.steps.') && consumers.length === 0) {
    const hits = blobs.filter((b) => /assessment/i.test(b.rel) && /steps/.test(b.text)).map((h) => h.rel);
    if (hits.length) {
      consumers.push(...hits);
      status = 'generated dynamically';
    }
  }
  return {
    key,
    consumers: [...new Set(consumers)].sort(),
    screen: screenFor(consumers),
    voice: voiceOf(preview, key),
    placeholders: placeholdersOf(preview) || '—',
    status,
    duplicateOf: duplicateOf(key, preview, nowFlat),
    preview: String(preview).replace(/\s+/g, ' ').slice(0, 180),
  };
});

const counts = {};
for (const r of rows) counts[r.status] = (counts[r.status] || 0) + 1;

const md = [];
md.push('# Phase 3 — added `ka.ts` key traceability');
md.push('');
md.push(`Generated: 2026-09-11`);
md.push(`HEAD: e518f9a`);
md.push(`HEAD keys: **${Object.keys(oldFlat).length}**`);
md.push(`Worktree keys: **${Object.keys(nowFlat).length}**`);
md.push(`Added keys: **${added.length}**`);
md.push(`Changed existing keys: **${changed.length}**`);
md.push('');
md.push('## Status counts');
for (const [k, v] of Object.entries(counts).sort()) md.push(`- ${k}: **${v}**`);
md.push(`- unclassified: **${added.length - rows.length}**`);
md.push('');
md.push('| Key | Consumer file(s) | Screen/state | Voice | Placeholder(s) | Status |');
md.push('| --- | ---------------- | ------------ | ----- | -------------- | ------ |');
for (const r of rows) {
  const cons = r.consumers.length ? r.consumers.map((c) => `\`${c}\``).join('<br>') : '—';
  const ph = r.placeholders === '—' ? '—' : `\`${r.placeholders}\``;
  md.push(`| \`${r.key}\` | ${cons} | ${r.screen} | ${r.voice} | ${ph} | ${r.status} |`);
}
md.push('');
md.push('## Unused added keys (not deleted)');
const unused = rows.filter((r) => r.status === 'unused');
md.push(`Count: **${unused.length}**`);
for (const r of unused) md.push(`- \`${r.key}\` — ${r.preview.slice(0, 100)}`);
md.push('');
md.push('## Duplicate text among added keys');
const dups = rows.filter((r) => r.duplicateOf);
md.push(`Count: **${dups.length}**`);
for (const r of dups) md.push(`- \`${r.key}\` same as \`${r.duplicateOf}\``);

fs.mkdirSync(OUT, { recursive: true });
fs.writeFileSync(path.join(OUT, 'ka-added-keys.md'), md.join('\n'), 'utf8');
fs.writeFileSync(
  path.join(OUT, 'ka-added-keys.json'),
  JSON.stringify(
    {
      head: 'e518f9a',
      head_key_count: Object.keys(oldFlat).length,
      worktree_key_count: Object.keys(nowFlat).length,
      added_count: added.length,
      changed_existing_count: changed.length,
      changed_existing: changed,
      status_counts: counts,
      rows,
    },
    null,
    2,
  ),
  'utf8',
);
console.log(
  JSON.stringify(
    {
      added: added.length,
      changed: changed.length,
      head: Object.keys(oldFlat).length,
      now: Object.keys(nowFlat).length,
      counts,
      unused: unused.length,
    },
    null,
    2,
  ),
);
console.log('carePlan sample', added.filter((k) => k.includes('carePlan')).slice(0, 5));
console.log('cycle sample', added.filter((k) => k.startsWith('cycle.')).slice(0, 15));
